### PR TITLE
3343 isy binding

### PR DESCRIPTION
**Pull request by mitchmitchell**
updated with formatting and other details to comply with review comments for merging isy binding into openhab1

### CODE REVIEW

**Comment by 9037568** [comment]
A couple of quick-hit notes while I dive into the rest of the code.

Copyright dates are showing "2016". These should be updated to "2017" throughout.
`@since` tags should be updated to use the value 1.10.0 throughout.

**Comment by mitchmitchell** [comment]
I'll fix those as soon as I get home...

**Comment by mitchmitchell** [comment]
ok a little sed scripting and the copyrights are updated and the since tags are fixed -- I've thing I have the updates properly added to the pull request, but perhaps not -- @9037568 chris can you take a look and see if you see the additional commits beyond the original 3 on my pull request?

**Comment by 9037568** [comment]
Thanks. 

a03e125 looks great.
796ddf7 made a lot of whitespace changes I wouldn't have expected, but the inheritdoc tags look improved.  I'll have to review the changes in Eclipse tomorrow.

I'm hoping to finish the full review soon also.

**Comment by mitchmitchell** [comment]
Thanks for the review, I will get to these tomorrow afternoon or this weekend.

File: distribution/openhabhome/configurations/openhab_default.cfg [hunk 1]
```diff
@@ -2734,3 +2734,26 @@ tcp:refreshinterval=250
...
+
+################################ ISY Binding #######################################
+#
+# Refresh interval in milliseconds (optional, defaults to 60000 [1 minute])
+# isy:refresh=60000 
+#
+# If true, use UPNP to communicate with the ISY 994i.
+# isy:upnp=false 
+#
+# UUID of the ISY router, git it fron the admin panel.
```
Comment (mitchmitchell) [inline distribution/openhabhome/configurations/openhab_default.cfg]: fixed

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/InsteonClient.java [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+
+                if (!authenticate(this.user, this.password)) {
+                    this.logger.warn("Authentication failed");
+
+                    return;
+                } else {
+                    this.logger.info("Authenticated");
+                }
+            } catch (NoDeviceException ne) {
+                this.logger.error("No device", ne);
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/InsteonClient.java]: changed to warning

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/InsteonClient.java [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+    public void onDeviceOnLine() {
+        this.logger.info("ISY is now online");
+
+        UDProxyDevice device = getDevice();
+
+        if (device == null) {
+            return;
+        }
+
+        //
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/InsteonClient.java]: removed

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ISYNodeType.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.isy.internal;
+
+/**
+ * This is the node types in ISY that are supported / translated.
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ISYNodeType.java]: changed

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                warning(status, null);
+                break;
+
+            case 801/* maximum secure sessions */:
+            case 805/* internal error: reboot */:
+            case 815/* Maximum Subscribers Reached */:
+            case 903/* System Not Initialized; Restart! */:
+            case 905/* Subscription Failed; The device might need reboot! */:
+            case 1020/* Couldn't create the event handler socket */:
+                fatalError(status, "Exit the applciation; might have to reboot ISY");
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: spelling fixed -- I don't know about the error handling, perhaps @sytone should comment, I am not that familiar with the algorithms used here.

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            UDControlPoint.getInstance().offLine((UDProxyDevice) device);
+        }
+
+    }
+
+    private void humanInterventionRequired(final int status, final String msg) {
+        String err = Errors.getErrorMessage(status);
+        this.logger.error("Human Intervention Required: " + status + (err == null ? " " : err));
+        if (msg != null) {
+            this.logger.error(msg);
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: changed to warning in both calls

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+        this.logger.warn("Warning: " + status + (err == null ? " " : err));
+        if (msg != null) {
+            this.logger.warn(msg);
+        }
+    }
+
+    private void fatalError(final int status, final String msg) {
+        String err = Errors.getErrorMessage(status);
+        this.logger.error("Fatal Error: " + status + (err == null ? " " : err));
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: I'm not sure what you mean by parameterized logging?  Do the logger apis accept varying number of parameters?

Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: error changed to warning

Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: shouldn't log messages in a routine called 'fatalError' be logged as errors?

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        }
+        if (device != null && device instanceof UDProxyDevice) {
+            UDControlPoint.getInstance().offLine((UDProxyDevice) device);
+        }
+
+    }
+
+    private void humanInterventionRequired(final int status, final String msg) {
+        String err = Errors.getErrorMessage(status);
+        this.logger.error("Human Intervention Required: " + status + (err == null ? " " : err));
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: changed to warning

Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: will research the logging apis this weekend

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+public class ErrorHandler implements ErrorEventListener {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private int socket_open_failed_count = 0;
+
+    private void offLine(final int status, Object device) {
+        String err = Errors.getErrorMessage(status);
+
+        this.logger.warn("Warning: " + status + (err == null ? " " : err));
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: will research logging apis this weekend

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYGenericBindingProvider.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                    address = value.replace('.', ' ');
+                    break;
+                case "type":
+                    if ("thermostat".equalsIgnoreCase(value)) {
+                        type = ISYNodeType.THERMOSTAT;
+                    }
+                    if ("lock".equalsIgnoreCase(value)) {
+                        type = ISYNodeType.LOCK;
+                    }
+                    if ("heartbeat".equalsIgnoreCase(value)) {
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYGenericBindingProvider.java]: fixed

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYGenericBindingProvider.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+                    controller = value.replace('.', ' ');
+                    break;
+                case "addr":
+                    address = value.replace('.', ' ');
+                    break;
+                case "type":
+                    if ("thermostat".equalsIgnoreCase(value)) {
+                        type = ISYNodeType.THERMOSTAT;
+                    }
+                    if ("lock".equalsIgnoreCase(value)) {
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYGenericBindingProvider.java]: fixed

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBindingConfig.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+            try {
+                this.port = Integer.parseInt((String) config.get("port"));
+            } catch (NumberFormatException ne) {
+                this.logger.error("Not a valid port number");
+                throw new RuntimeException("Not a valid port number", ne);
+            }
+
+            if (this.port < 0 || this.port > 65535) {
+                this.logger.error("Port out of range (0-65535)");
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBindingConfig.java]: changed to warning

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBindingConfig.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        if (!useUpnp && isNotBlank((String) config.get("uuid")) && isNotBlank((String) config.get("ip"))
+                && isNotBlank((String) config.get("port"))) {
+
+            this.uuid = (String) config.get("uuid");
+            this.ipAddress = (String) config.get("ip");
+
+            try {
+                this.port = Integer.parseInt((String) config.get("port"));
+            } catch (NumberFormatException ne) {
+                this.logger.error("Not a valid port number");
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBindingConfig.java]: changed to warning

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java [hunk 14]
```diff
@@ -0,0 +1,680 @@
...
+                }
+                break;
+
+            default:
+                this.logger.debug("Unsupported control '{}'", control.name);
+        }
+    }
+
+    /**
+     * This will look though items that have a binding to the ISY unit and
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java]: spelling correct.

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java [hunk 15]
```diff
@@ -0,0 +1,680 @@
...
+    @Override
+    public void deactivate() {
+        if (this.insteonClient != null) {
+            this.insteonClient.stop();
+            this.insteonClient = null;
+        }
+    }
+
+    /**
+     * @{inheritDoc
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java]: fixed here, will search and fix elsewhere.

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java [hunk 16]
```diff
@@ -0,0 +1,680 @@
...
+    }
+
+    /**
+     * Called by SCR when the binding configuration was modified
+     */
+    public void modified(final Map<Object, Object> config) {
+
+        runtimeConfiguration = new ISYActiveBindingConfig(config);
+
+        setProperlyConfigured(true);
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java]: moved as requested.

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java [hunk 17]
```diff
@@ -0,0 +1,680 @@
...
+    /**
+     * Called by SCR when the binding is resolved
+     *
+     * @param config
+     */
+    public void activate(final Map<Object, Object> config) {
+
+        runtimeConfiguration = new ISYActiveBindingConfig(config);
+
+        setProperlyConfigured(true);
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java]: moved as requested.

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java [hunk 18]
```diff
@@ -0,0 +1,680 @@
...
+
+    private ISYActiveBindingConfig runtimeConfiguration;
+
+    /**
+     * The client used to send and receive messages from ISY.
+     */
+    private InsteonClient insteonClient;
+
+    /**
+     * Constructor, nothing to see here.
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java]: removed.

File: bundles/binding/org.openhab.binding.isy/META-INF/MANIFEST.MF [hunk 19]
```diff
@@ -0,0 +1,30 @@
+Manifest-Version: 1.0
+Private-Package: org.openhab.binding.isy.internal
+Ignore-Package: org.openhab.binding.isy.internal
+Bundle-License: http://www.eclipse.org/legal/epl-v10.html
+Bundle-Name: openHAB isy Binding
+Bundle-SymbolicName: org.openhab.binding.isy
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.9.0.qualifier
```
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/META-INF/MANIFEST.MF]: changed.

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java (hunk 15 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/ISYActiveBinding.java]: fixed in all places, fgrep -R confirmed...

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 9 above)
Comment (9037568) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: Here's what this call would look like using parameterized logging:
>this.logger.warn("Warning: {}{}", status, err == null ? " " : err);

Thanks for all the fixes!

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 7 above)
Comment (9037568) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: > shouldn't log messages in a routine called 'fatalError' be logged as errors?

Right. The [logging guidelines ](http://docs.openhab.org/developers/development/guidelines.html#e-logging) are basically that 'error' is for tremendous setup errors or things that would qualify as severe bugs.

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 9 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: changed to parameterized logging and joined msg parameter to make only one call

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 8 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: changed to parameterized logging and joined msg parameter to make only one call

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 6 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: changed to parameterized logging and joined msg parameter to make only one call

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 7 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: changed to warning added verbiage to indicate that the ISY itself may have suffered a failure.

**Comment by mitchmitchell** [comment]
ok I made the remaining changes -- I built the jar file so things compile but my openhab is now on the 2.1 nightly build so I will need someone to test it on an openhab1 system.

**Comment by mitchmitchell** [comment]
ok I pushed the last of the changes from your review, please let me know if I need to do anything else!

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 8 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: fixed

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 9 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: fixed

File: bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java (hunk 7 above)
Comment (mitchmitchell) [inline bundles/binding/org.openhab.binding.isy/src/main/java/org/openhab/binding/isy/internal/ErrorHandler.java]: fixed

**Comment by 9037568** [comment]
Thanks, @mitchmitchell ! Much appreciated.

**Comment by mitchmitchell** [comment]
Thank you!!

**Pull request merged**